JSON: Add support for Int128, UInt128 and Half

jozkee opened the pull request:
~~...and add Number support for `Utf8JsonReader.CopyString(...)`.~~
EDIT: Instead of changing CopyString to accept numbers, we will just enable it on an internal helper since we could regret the decission.

Fixes https://github.com/dotnet/runtime/issues/87994

jozkee committed:
JSON: Add support for Int128, UInt128 and Half and add Number support for Utf8JsonReader.CopyString(...)

Jozkee commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,146 @@
+{
+    internal sealed class HalfConverter : JsonPrimitiveConverter<Half>
+    {
+        private const int MaxFormatLength = 16;
+        private const int MaxEscapedFormatLength = MaxFormatLength * JsonConstants.MaxExpansionFactorWhileEscaping;
```
I'm not entirely what's a good number here since I couldn't find a good way to determine what's the max amount of bytes `Half.TryFormat` can use.

eiriktsarpalis replied:
Presumably you mean `Half.TryParse`? @tannergooding might know

tannergooding replied:
What are you trying to do here in particular? The underlying parsing algorithm needs to be able to track up to 20 significant digits (for `Half`, it's 113 for `Single`, and 768 for `Double`) to ensure a correct result.

However, the entire input string must always be passed in such that it can process non-significant digits (such as leading zeros) and all trailing digits so that it can determine if the rounding goes up or down.

Imagine for example if the user defines `000...005` or `0.500...1`, etc. All the zero digits that represented by the `...` must be processed to ensure the result is correct and to ensure the relevant end of string is located.

Jozkee replied:
Actually I suspect BitConverter.ToHalf doesn't have the TryParse wiggle room.

tannergooding replied:
That's not an equivalent API. `BitConverter.ToHalf` simply does a reinterpret cast of raw bytes into a `Half`.

You'd need to do something similar to the default interface implementation for `IUtf8SpanParsable` done by `INumberBase<T>` here: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Numerics/INumberBase.cs,555

Which is to say, you have to transcode the input string to UTF-16, then try to parse.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,146 @@
+
+            if (reader.TokenType is JsonTokenType.String or JsonTokenType.PropertyName)
+            {
+                if (JsonReaderHelper.TryGetFloatingPointConstant(buffer, out result))
```
Isn't it the case that we only try to read floating point literals if `JsonNumberHandling.AllowNamedFloatingPointLiterals` is enabled? This appears to be the default deserialization path.

Jozkee replied:
No, if you look at `SingleConverter`:
https://github.com/dotnet/runtime/blob/1380a0b327a24a924ef3f63f144558c742da0a3c/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/SingleConverter.cs#L37-L49

When `AllowReadingFromString` is specified, we call GetSingleWithQuotes() which checks both floating point literals and numbers.

I agree it sounds counterintuitive because JsonNumberHandling is a Flags enum, but that's how it is.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,146 @@
+#if NET8_0_OR_GREATER
+            bool success = Half.TryParse(buffer, out result);
+#else
+            Span<char> charBuffer = stackalloc char[MaxFormatLength];
```
I wonder if much of the complexity in this converter might be eliminated if we simply converted the `Half` value to `float` or `double` when serializing and vice versa when deserializing, using the existing implementation for these types. 

@tannergooding might be able to clarify if there are any gotchas in such an approach.

tannergooding replied:
That could potentially lead to incorrect results due to double rounding caused by the different domain boundaries. Someone would need to sit down and do the math to determine if it was definitively safe or not, and that's likely significantly more effort.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/Int128Converter.cs:
```diff
@@ -0,0 +1,120 @@
+#else
+            Span<char> buffer = stackalloc char[MaxFormatLength];
+#endif
+            int written = reader.CopyString(buffer);
```
Given that you're reading a number token that can't contained escaping, I think it should be safe to just pass `reader.ValueSpan` to the `Int128` parsing method.

eiriktsarpalis replied:
Ah, I see now that you're using the same method for the property name parsing logic. Should be fine I suppose.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/Int128Converter.cs:
```diff
@@ -0,0 +1,120 @@
+#if NET8_0_OR_GREATER
+            Span<byte> buffer = stackalloc byte[MaxFormatLength];
+#else
+            Span<char> buffer = stackalloc char[MaxFormatLength];
```
Is this because UTF-8 parsing overloads don't exist in .NET 7 presumably?

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/Common/JsonTestHelper.cs:
```diff
@@ -20,6 +20,16 @@ internal static partial class JsonTestHelper
+#if NETCOREAPP
+        public static Half NextHalf(Random random)
+        {
+            double mantissa = (random.NextDouble() * 2.0) - 1.0;
```
Nit: are you looking to avoid creating values like `NaN` and infinity here? If not just generating a random pair of bytes and passing them to BitConverter.ToHalf might be simpler.

eiriktsarpalis replied:
Nevermind, I just realised you're following existing convention for other types.

Jozkee replied:
> are you looking to avoid creating values like NaN and infinity here?

Yes, that was the reason why I did this method after trying with BitConverter and the random pair of bytes.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/Common/JsonNumberTestData.cs:
```diff
@@ -237,6 +247,58 @@ static JsonNumberTestData()
+                0,
+                (Int128)long.MaxValue + 1,
+                (Int128)long.MinValue - 1,
+                Int128.MaxValue,
```
Do we have any tests checking what happens when you try to deserialize a JSON number equalling `Int128.MaxValue + 1` and `Int128.MinValue - 1`?

eiriktsarpalis replied:
We would need some kind of testing that validates handling of payloads with unsupported JSON numbers such as numbers exceeding `Int128.MaxValue` and floating point numbers.

krwq commented on src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs:
```diff
@@ -68,11 +68,13 @@ namespace System.Text.Json
         public readonly int CopyString(Span<byte> utf8Destination)
         {
-            if (_tokenType is not (JsonTokenType.String or JsonTokenType.PropertyName))
+            if (_tokenType is not (JsonTokenType.String or JsonTokenType.PropertyName or JsonTokenType.Number))
```
Is this actually what we want? It doesn't seem particularly intuitive that GetString works for numbers...

If this is what we want we should update the xml doc

eiriktsarpalis replied:
I see your point, in hindisight this should probably have been named `CopyValue` to avoid confusion. My rationalization for this change is that 1) it unblocks a useful scenario and 2) JSON numbers are technically string representations without quote marks.

> If this is what we want we should update the xml doc

Yes, absolutely 👍

Jozkee replied:
@krwq I got the same concern while working on it. If we want to avoid making a change that we then will regret, we could just make an `internal CopyValue(Span<byte>)` helper. 

I don't have a strong opinion, though.

eiriktsarpalis replied:
I'm ok with using an internal helper for now.

jozkee committed:
Remove parsing limits on Read and move Number support of CopyString to an internal helper

jozkee committed:
Fix AllowNamedFloatingPointLiterals on Write for Half

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -34,36 +34,57 @@ private static Half ReadCore(ref Utf8JsonReader reader)
+
+                return result;
             }
+            finally
```
We generally don't use try/finally to guard against exceptions in code that uses rented buffers. Not returning a buffer because of an exception is not a big problem all things considered (the problems start if a buffer gets used after being returns or gets returned more than once). You should still move the returning logic above the `if (!success)` statement above though.

eiriktsarpalis reviewed (approved):
Other than a few pending issues that should be addressed this looks good to me. Great work David!

jozkee committed:
Specify InvariantCulture on TryParse and TryFormat
Fix handling of floating-point literals on HalfConverter
Remove CopyString tests related to Number support

jozkee committed:
Add test for invalid number input format

Jozkee commented:
@eiriktsarpalis can you please take another look at the last commits, I found a couple of issues:
* For formating infinites we were writing them as `∞` and `-∞`, specifying `CultureInfo.InvariantCulture` fixed it.
* For parsing NaN and infinites, Half.TryParse was more lax than the current S.T.Json policy and accepted them with any case i.e: `InFiNiTy` could be parsed correctly; I fixed it by comparing the bytes when the TryParse method returned NaN, PositiveInfinity or NegativeInfinity.

jozkee committed:
Fix net6.0 build error about missing Half.TryParse overload

eiriktsarpalis commented:
> For formating infinites we were writing them as ∞ and -∞, specifying CultureInfo.InvariantCulture fixed it.

What configuration is being used when the corresponding values for `double` and `float` are being used?

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/Common/NumberHandlingTests.cs:
```diff
@@ -1057,6 +1057,27 @@ private async Task AssertUnsupportedFormatThrows<T>()
+        [Fact]
+        public async Task InvalidNumberFormatThrows()
+        {
+            await AssertInvalidNumberFormatThrows<Int128>("170141183460469231731687303715884105728"); // MaxValue + 1
```
Nit: could this be a theory?

Jozkee replied:
Yes it could, but this is pattern used by the nearby tests in this file.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/Resources/Strings.resx:
```diff
@@ -690,4 +690,13 @@
   <data name="ObjectCreationHandlingPropertyCannotAllowReferenceHandling" xml:space="preserve">
     <value>JsonObjectCreationHandling.Populate is incompatible with reference handling.</value>
   </data>
+  <data name="FormatInt128" xml:space="preserve">
```
I think you could collapse all these messages to a single string that accepts the type name as a parameter.

Jozkee replied:
Yes that is true, but we should do it for all the `Format*` strings.
https://github.com/dotnet/runtime/blob/f7ad7265b222bede49ff93b341d81b61bd898122/src/libraries/System.Text.Json/src/Resources/Strings.resx#L265-L285

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,228 @@
+        // Half.TryFormat/TryParse(ROS<byte>) are not available on .NET 7
+        // we need to use Half.TryFormat/TryParse(ROS<char>) in that case.
+        private static bool TryParse(
+#if NET8_0_OR_GREATER
```
Clarifying a bit more what I meant in https://github.com/dotnet/runtime/pull/88962/files#r1266760827, I think `TryParse` could be made to only accept a `ROS<byte>` argument and only transcode inside the method if `!NET8_0_OR_GREATER`.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,228 @@
+            bool success = Half.TryParse(buffer, NumberStyles.Float | NumberStyles.AllowThousands, CultureInfo.InvariantCulture, out result);
+
+            // Half.TryParse is more lax with floating-point literals than other S.T.Json floating-point types
+            // e.g: it parses "naN" successfully. Only succeed with the exact match.
```
If I'm understanding this correctly, Trying to deserialize `{ "number" : NaN }` will be successful whereas `{ "number" : naN }` will fail.

However both cases are invalid JSON since non-numeric identifiers can only be contained in JSON strings.

Given that this method will get called in cases where you have `JsonTokenType.Number`, it should always reject non-numeric identifiers.

Jozkee replied:
> Trying to deserialize { "number" : NaN } will be successful whereas { "number" : naN } will fail.

No, they need be inside quotes, otherwise the reader would throw invalid token:

```
var reader = new Utf8JsonReader(Encoding.UTF8.GetBytes("NaN"));
Console.WriteLine(reader.Read());
// throws: Unhandled exception. System.Text.Json.JsonReaderException: 'N' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
```

Jozkee replied:
> Given that this method will get called in cases where you have JsonTokenType.Number

It can actually be called when TokenType is JsonTokenType.String or JsonTokenType.PropertyName too when `ReadCore` is called from `ReadAsPropertyNameCore` or `ReadNumberWithCustomHandling`.

Jozkee commented:
> What configuration is being used when the corresponding values for double and float are being used?

InvariantCulture as well, that's the default for Utf8Formatter.TryFormat:
https://github.com/dotnet/runtime/blob/f7ad7265b222bede49ff93b341d81b61bd898122/src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs#L19
`formatText` is also equivalent to what we use for the new Converters.

Jozkee commented:
One of our tests found an assertion failure on `Half.TryParse` that has reproed consistently only on OSX.

https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-88962-merge-9757e201f6374e20b5/System.Text.Json.Tests/1/console.78feb75a.log?helixlogtype=result

```
/private/tmp/helix/working/9EC208C7/w/A9F108F7/e /private/tmp/helix/working/9EC208C7/w/A9F108F7/e
  Discovering: System.Text.Json.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Text.Json.Tests (found 7160 of 7231 test cases)
  Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 6)
Process terminated. Assertion failed.
   at System.Globalization.Ordinal.EqualsIgnoreCaseUtf8_Scalar(Byte& charA, Int32 lengthA, Byte& charB, Int32 lengthB) in /_/src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.Utf8.cs:line 308
   at System.Number.TryParseFloat[TChar,TFloat](ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info, TFloat& result) in /_/src/libraries/System.Private.CoreLib/src/System/Number.Parsing.cs:line 1229
   at System.Half.TryParse(ReadOnlySpan`1 utf8Text, NumberStyles style, IFormatProvider provider, Half& result) in /_/src/libraries/System.Private.CoreLib/src/System/Half.cs:line 2238
   at System.Text.Json.Serialization.Converters.HalfConverter.TryParse(ReadOnlySpan`1 buffer, Half& result) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:line 197
   at System.Text.Json.Serialization.Converters.HalfConverter.ReadCore(Utf8JsonReader& reader) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:line 50
   at System.Text.Json.Serialization.Converters.HalfConverter.ReadNumberWithCustomHandling(Utf8JsonReader& reader, JsonNumberHandling handling, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:line 115
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:line 193
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs:line 308
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectDefaultConverter.cs:line 49
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs:line 258
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.ReadCore.cs:line 51
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.Deserialize(Utf8JsonReader& reader, ReadStack& state) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.ReadHelper.cs:line 22
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo`1 jsonTypeInfo, Nullable`1 actualByteCount) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Span.cs:line 160
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo`1 jsonTypeInfo) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs:line 443
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs:line 55
   at System.Text.Json.Serialization.Tests.JsonSerializerWrapper.StringSerializerWrapper.DeserializeWrapper[T](String json, JsonSerializerOptions options) in /_/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/JsonSerializerWrapper.Reflection.cs:line 129
   at System.Text.Json.Serialization.Tests.NumberHandlingTests.<>c__DisplayClass39_0.<<FloatingPointConstants_Fail>b__1>d.MoveNext() in /_/src/libraries/System.Text.Json/tests/Common/NumberHandlingTests.cs:line 883
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilderCore.cs:line 38
   at System.Text.Json.Serialization.Tests.NumberHandlingTests.<>c__DisplayClass39_0.<FloatingPointConstants_Fail>b__1()
```

jozkee committed:
Move rentedCharBuffer logic to TryParse helper

Jozkee commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,228 @@
+            ReadOnlySpan<byte> PositiveInfinity = JsonConstants.PositiveInfinityValue;
+            ReadOnlySpan<byte> NegativeInfinity = JsonConstants.NegativeInfinityValue;
+#else
+            const string NaN = "NaN";
```
I was able to remove these `const string`s in the latest commit, since we are passing the bytes in all tfms, we can just compare with the existing utf8 JsonConstants.

eiriktsarpalis commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs:
```diff
@@ -0,0 +1,217 @@
+            }
+        }
+
+        // Half.TryFormat/TryParse(ROS<byte>) are not available on .NET 7
```
Nit: move the comment inside the `#else` clause?

Jozkee replied:
I can do that, but I don't want to retrigger CI just for this 😅.

eiriktsarpalis commented on src/libraries/System.Text.Json/tests/Common/NumberHandlingTests.cs:
```diff
@@ -985,6 +1057,27 @@ private async Task AssertUnsupportedFormatThrows<T>()
+            await AssertInvalidNumberFormatThrows<Int128>("3.14");
+            await AssertInvalidNumberFormatThrows<UInt128>("-1");
+            await AssertInvalidNumberFormatThrows<Half>("65520");
+            await AssertInvalidNumberFormatThrows<Half>("-65520");
```
Per @GrabYourPitchforks feedback could you add a test case that uses leading zeros?

Jozkee replied:
This might be a better place to add the test case:
https://github.com/dotnet/runtime/blob/43fe8a8fc7642a1268e80b0433aebb92b049ee09/src/libraries/System.Text.Json/tests/Common/NumberHandlingTests.cs#L1044C28-L1052

jozkee committed:
Address feedback

jozkee committed:
Disable test for OSX
